File: packages/protocol/src/canonical.ts
/** Deterministic JSON for Ed25519 signing (sorted object keys, stable arrays). */

export function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => stableStringify(item)).join(",")}]`;
  }
  const record = value as Record<string, unknown>;
  const keys = Object.keys(record)
    .filter((key) => record[key] !== undefined)
    .sort();
  return `{${keys.map((key) => `${JSON.stringify(key)}:${stableStringify(record[key])}`).join(",")}}`;
}

/** Fields covered by the envelope signature. */
export function signingPayload(object: {
  version: 1;
  id: string;
  issuerDid: string;
  issuedAt: string;
  semantic: unknown;
  payload: unknown;
  governance: unknown;
}): string {
  return stableStringify({
    version: object.version,
    id: object.id,
    issuerDid: object.issuerDid,
    issuedAt: object.issuedAt,
    semantic: object.semantic,
    payload: object.payload,
    governance: object.governance,
  });
}

export function bytesToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary);
}

export function base64ToBytes(encoded: string): Uint8Array {
  const binary = atob(encoded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

File: spec/second-impl/atom_gov/canonical.py
"""Deterministic JSON for Ed25519 signing (sorted object keys, stable arrays)."""

from __future__ import annotations

import base64
import json
from typing import Any


def stable_stringify(value: Any) -> str:
    if value is None or isinstance(value, (bool, int, float, str)):
        return json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if isinstance(value, list):
        return "[" + ",".join(stable_stringify(item) for item in value) + "]"
    if isinstance(value, dict):
        # Keep JSON null (Python None); omit only absent keys — mirrors JS undefined filter.
        keys = sorted(value.keys())
        inner = ",".join(
            f"{json.dumps(k, ensure_ascii=False)}:{stable_stringify(value[k])}"
            for k in keys
        )
        return "{" + inner + "}"
    raise TypeError(f"unsupported type for stable_stringify: {type(value)!r}")


def signing_payload(obj: dict[str, Any]) -> str:
    return stable_stringify(
        {
            "version": obj["version"],
            "id": obj["id"],
            "issuerDid": obj["issuerDid"],
            "issuedAt": obj["issuedAt"],
            "semantic": obj["semantic"],
            "payload": obj["payload"],
            "governance": obj["governance"],
        }
    )


def b64_decode(encoded: str) -> bytes:
    return base64.b64decode(encoded, validate=True)
